zebrad/src/components/sync/status.rs

```diff
@@ -28,11 +67,36 @@ impl SyncStatus {
 
     /// Create an instance of [`SyncStatus`].
     ///
-    /// The status is determined based on the latest counts of synchronized blocks, observed
-    /// through `latest_sync_length`.
-    pub fn new() -> (Self, RecentSyncLengths) {
-        let (recent_sync_lengths, latest_sync_length) = RecentSyncLengths::new();
-        let status = SyncStatus { latest_sync_length };
+    /// - `peer_status` feeds live peer availability so we can gate readiness on connectivity.
+    /// - `latest_chain_tip` lets us observe progress that bypasses the syncer (e.g. gossip).
+    /// - `min_ready_peers` encodes how many ready peers we require before enabling tip logic.
+    /// - `progress_grace` matches the health config and bounds how long we treat the last
+    ///   observed activity as "recent" when sync batches are zero.
+    pub fn new(
+        peer_status: watch::Receiver<PeerSetStatus>,
+        latest_chain_tip: zs::LatestChainTip,
+        min_ready_peers: usize,
+        progress_grace: Duration,
+    ) -> (Self, RecentSyncLengths) {
+        let tip_progress = Arc::new(Mutex::new(TipProgress::new()));
+
+        let progress_tip = Arc::clone(&tip_progress);
+        let progress_callback: Arc<dyn Fn() + Send + Sync> = Arc::new(move || {
+            let mut tip = progress_tip.lock().expect("tip progress mutex poisoned");
+            tip.last_updated = Instant::now();
+            tip.has_progress = true;
+        });
```

Comment:
We generally prefer watch channels over `Mutex` where possible, in this case, the `last_updated` time and `has_progress` flag can be added as fields to the `RecentSyncLength`'s watch channel message type, though I'm not sure it's necessary to record those at all.